confluent_kafka/consumer.py

```diff
@@ -0,0 +1,145 @@
+                 on_commit=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_deserializer = key_deserializer
+        self._value_deserializer = value_deserializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
+        # If callbacks are configured in both places parameter values take precedence.
+        if not on_commit:
+            on_commit = conf.get('on_commit', None)
+
+        if not stats_cb:
+            stats_cb = conf.get('stats_cb', None)
+
+        if not throttle_cb:
+            throttle_cb = conf.get('throttle_cb', None)
+
+        if not logger:
+            logger = conf.get('logger', None)
+        super(Consumer, self).__init__(conf, on_commit=on_commit, stats_cb=stats_cb,
+                                       throttle_cb=throttle_cb, logger=logger)
+
+    def poll(self, timeout=-1.0, key_deserializer=None, value_deserializer=None):
+        """
+            Consumes a message, triggers callbacks, returns an event.
+
+            The application must check the returned Message object’s Message.error() method to distinguish
+            between proper messages, an error(see error().code() for specifics), or an event.
+
+            :param float timeout:  Maximum time in seconds to block waiting for message, event or callback.
+            :param func key_deserializer(topic, key): Converts message key bytes to object.
+                **note** deserializers are responsible for handling NULL keys
+            :param func value_deserializer(topic, value): Converts message value bytes to object.
+                **note** deserializers are responsible for handling NULL values
+            :returns: A confluent_kafka.Message or None on timeout.
+            :raises RuntimeError: If called on a closed consumer.
+        """
+
+        msg = super(Consumer, self).poll(timeout)
+
+        if not msg or msg.error():
+            return msg
+
+        topic = msg.topic()
+
+        # parameter overrides take precedence over instance functions
+        if not key_deserializer:
+            key_deserializer = self._key_deserializer
+
+        if key_deserializer:
+            msg.set_key(key_deserializer(topic, msg.key()))
+
+        if not value_deserializer:
+            value_deserializer = self._value_deserializer
+
+        if value_deserializer:
+            msg.set_value(value_deserializer(topic, msg.value()))
+
```

Comment:
what are the performance implications of this wrapping and extra checks? (for both consumer and producer)
I want numbers!

Reply:
Not sure about the details, but it seems like we might ideally want a super fast path for (at least) `UTF8`, `int` and `long` where the deserialization happens in C. Have you investigated a three-pronged approach (no/fast/custom deserialization) along these lines? I think we might get string for free with python 2, but not python 3... not an expert though and haven't investigated.

Reply:
All Python objects are of type PyObject and will need to be converted to their c counterparts so I'm not sure I understand what you mean here. Even bytes are of type PyBytes in Py3 so I'm not sure you get anything free in either version. 

https://docs.python.org/2/c-api/string.html